File: Polars.CSharp/IO/IPC/Read.cs
#pragma warning disable CS1573
using Polars.NET.Core;
using Cs = Polars.CSharp.Polars.Selectors;

namespace Polars.CSharp;
public partial class DataFrame : IDisposable,IEnumerable<Series>,IPolarsDataFrame
{
    // ---------------------------------------------------------
    // Read IPC (File / Cloud)
    // ---------------------------------------------------------

    /// <summary>
    /// Read an Arrow IPC (Feather v2) file into a DataFrame.
    /// <para>
    /// Note: This method internally uses LazyFrame.ScanIpc and collects the result. 
    /// For larger-than-memory datasets or better query optimization, consider using LazyFrame.ScanIpc directly.
    /// </para>
    /// </summary>
    /// <param name="path">Path to the IPC file, glob pattern, or cloud path (e.g., "s3://...").</param>
    /// <param name="columns">Columns to select. If null, select all columns. (Optimized via projection pushdown).</param>
    /// <param name="schema">Optional schema to enforce. If not provided, the schema is inferred from the file footer.</param>
    /// <param name="nRows">Limit the number of rows to scan.</param>
    /// <param name="rechunk">Make sure the DataFrame is contiguous in memory (default: false).</param>
    /// <param name="cache">Cache the result of the scan (default: true).</param>
    /// <param name="glob">Expand glob patterns (default: true).</param>
    /// <param name="rowIndexName">If provided, adds a column with the row index.</param>
    /// <param name="rowIndexOffset">Offset for the row index (default: 0).</param>
    /// <param name="includePathColumn">If provided, adds a column with the source file path.</param>
    /// <param name="hivePartitioning">Enable Hive partitioning inference (default: false).</param>
    /// <param name="hivePartitionSchema">Manually specify the schema for Hive partitioning columns.</param>
    /// <param name="tryParseHiveDates">Whether to try parsing dates in Hive partitioning paths (default: true).</param>
    /// <param name="cloudOptions">Options for cloud storage (AWS S3, Azure Blob, GCS, etc.).</param>
    public static DataFrame ReadIpc(
        string path,
        string[]? columns = null,
        PolarsSchema? schema = null,
        ulong? nRows = null,
        bool rechunk = false,
        bool cache = true,
        bool glob = true,
        string? rowIndexName = null,
        uint rowIndexOffset = 0,
        string? includePathColumn = null,
        bool hivePartitioning = false,
        PolarsSchema? hivePartitionSchema = null,
        bool tryParseHiveDates = true,
        CloudOptions? cloudOptions = null)
    {
        var lf = LazyFrame.ScanIpc(
            path,
            schema,
            nRows,
            rechunk,
            cache,
            glob,
            rowIndexName,
            rowIndexOffset,
            includePathColumn,
            hivePartitioning,
            hivePartitionSchema,
            tryParseHiveDates,
            cloudOptions
        );

        if (columns != null && columns.Length > 0)
        {
            var colsToSelect = new List<string>(columns);
            
            if (!string.IsNullOrEmpty(rowIndexName) && !colsToSelect.Contains(rowIndexName))
            {
                colsToSelect.Add(rowIndexName);
            }

            if (!string.IsNullOrEmpty(includePathColumn) && !colsToSelect.Contains(includePathColumn))
            {
                colsToSelect.Add(includePathColumn);
            }

            lf = lf.Select(Cs.ByName([.. colsToSelect]));
        }

        return lf.Collect();
    }

    // ---------------------------------------------------------
    // Read IPC (Memory / Bytes)
    // ---------------------------------------------------------

    /// <summary>
    /// Read Arrow IPC (Feather v2) from in-memory bytes.
    /// <para>
    /// Note: This method internally uses LazyFrame.ScanIpc and collects the result.
    /// </para>
    /// </summary>
    /// <param name="buffer">The byte array containing the IPC data.</param>
    /// <param name="columns">Columns to select. If null, select all columns.</param>
    /// <inheritdoc cref="ReadIpc(string, string[], PolarsSchema?, ulong?, bool, bool, bool, string?, uint, string?, bool, PolarsSchema?, bool, CloudOptions?)"/>
    public static DataFrame ReadIpc(
        byte[] buffer,
        string[]? columns = null,
        PolarsSchema? schema = null,
        ulong? nRows = null,
        bool rechunk = false,
        bool cache = true,
        string? rowIndexName = null,
        uint rowIndexOffset = 0,
        string? includePathColumn = null,
        bool hivePartitioning = false,
        PolarsSchema? hivePartitionSchema = null,
        bool tryParseHiveDates = false)
    {
        var lf = LazyFrame.ScanIpc(
            buffer,
            schema,
            nRows,
            rechunk,
            cache,
            rowIndexName,
            rowIndexOffset,
            includePathColumn,
            hivePartitioning,
            hivePartitionSchema,
            tryParseHiveDates
        );

        if (columns != null && columns.Length > 0)
        {
            var colsToSelect = new List<string>(columns);
            
            if (!string.IsNullOrEmpty(rowIndexName) && !colsToSelect.Contains(rowIndexName))
            {
                colsToSelect.Add(rowIndexName);
            }

            if (!string.IsNullOrEmpty(includePathColumn) && !colsToSelect.Contains(includePathColumn))
            {
                colsToSelect.Add(includePathColumn);
            }

            lf = lf.Select(Cs.ByName([.. colsToSelect]));
        }

        return lf.Collect();
    }

    // ---------------------------------------------------------
    // Read IPC (Stream)
    // ---------------------------------------------------------

    /// <summary>
    /// Read Arrow IPC (Feather v2) from a Stream.
    /// <para>
    /// Note: This reads the stream fully into memory and then uses the Lazy execution engine.
    /// </para>
    /// </summary>
    /// <param name="stream">The input stream containing the IPC data.</param>
    /// <param name="columns">Columns to select. If null, select all columns.</param>
    /// <inheritdoc cref="ReadIpc(string, string[], PolarsSchema?, ulong?, bool, bool, bool, string?, uint, string?, bool, PolarsSchema?, bool, CloudOptions?)"/>
    public static DataFrame ReadIpc(
        Stream stream,
        string[]? columns = null,
        PolarsSchema? schema = null,
        ulong? nRows = null,
        bool rechunk = false,
        bool cache = true,
        string? rowIndexName = null,
        uint rowIndexOffset = 0,
        string? includePathColumn = null,
        bool hivePartitioning = false,
        PolarsSchema? hivePartitionSchema = null,
        bool tryParseHiveDates = false)
    {
        using var ms = new MemoryStream();
        stream.CopyTo(ms);
        
        return ReadIpc(
            ms.ToArray(),
            columns,
            schema,
            nRows,
            rechunk,
            cache,
            rowIndexName,
            rowIndexOffset,
            includePathColumn,
            hivePartitioning,
            hivePartitionSchema,
            tryParseHiveDates
        );
    }
}

File: Polars.FSharp/IO/IPC/Read.fs
namespace Polars.FSharp

open Polars.NET.Core

[<AutoOpen>]
module IPCRead =
    open System.IO
    type LazyFrame with
        /// <summary>
        /// Lazily read an Arrow IPC (Feather v2) file, multiple files via glob patterns, or cloud storage.
        /// </summary>
        static member ScanIpc(
            path: string,
            ?schema: PolarsSchema,
            ?nRows: uint64,
            ?rechunk: bool,
            ?cache: bool,
            ?glob: bool,
            ?rowIndexName: string,
            ?rowIndexOffset: uint32,
            ?includePathColumn: string,
            ?hivePartitioning: bool,
            ?hivePartitionSchema: PolarsSchema,
            ?tryParseHiveDates: bool,
            ?cloudOptions: CloudOptions
        ) : LazyFrame =
            
            // Resolve Optional Handles
            let schemaHandle = match schema with Some s -> s.Handle | None -> null
            let hiveSchemaHandle = match hivePartitionSchema with Some s -> s.Handle | None -> null
            
            // Resolve Defaults
            let rows = Option.toNullable nRows
            let rechk = defaultArg rechunk false
            let useCache = defaultArg cache true 
            let useGlob = defaultArg glob true
            let idxName = Option.toObj rowIndexName
            let idxOffset = defaultArg rowIndexOffset 0u
            let pathCol = Option.toObj includePathColumn
            let hive = defaultArg hivePartitioning false
            let hiveDates = defaultArg tryParseHiveDates true

            // Parse Cloud Options
            let cProv, cRet, cToMs, cInitMs, cMaxMs, cCache, cKeys, cVals = 
                CloudOptions.ParseCloudOptions cloudOptions

            let h = PolarsWrapper.ScanIpc(
                path, 
                rows, 
                rechk, 
                useCache, 
                useGlob,
                idxName, 
                idxOffset, 
                pathCol, 
                schemaHandle,
                hive,
                hiveSchemaHandle,
                hiveDates,
                // Cloud
                cProv,
                cRet,
                cToMs,
                cInitMs,
                cMaxMs,
                cCache,
                cKeys,
                cVals
            )
            new LazyFrame(h)

        // ---------------------------------------------------------
        // Scan IPC (Memory / Bytes)
        // ---------------------------------------------------------

        /// <summary>
        /// Lazily read Arrow IPC (Feather v2) from in-memory bytes.
        /// </summary>
        static member ScanIpc(
            buffer: byte[],
            ?schema: PolarsSchema,
            ?nRows: uint64,
            ?rechunk: bool,
            ?cache: bool,
            ?rowIndexName: string,
            ?rowIndexOffset: uint32,
            ?includePathColumn: string,
            ?hivePartitioning: bool,
            ?hivePartitionSchema: PolarsSchema,
            ?tryParseHiveDates: bool
        ) : LazyFrame =
            
            let schemaHandle = match schema with Some s -> s.Handle | None -> null
            let hiveSchemaHandle = match hivePartitionSchema with Some s -> s.Handle | None -> null
            
            let rows = Option.toNullable nRows
            let rechk = defaultArg rechunk false
            let useCache = defaultArg cache true
            let idxName = Option.toObj rowIndexName
            let idxOffset = defaultArg rowIndexOffset 0u
            let pathCol = Option.toObj includePathColumn
            let hive = defaultArg hivePartitioning false
            let hiveDates = defaultArg tryParseHiveDates false

            let h = PolarsWrapper.ScanIpc(
                buffer, 
                rows, 
                rechk, 
                useCache, 
                idxName, 
                idxOffset, 
                pathCol,
                schemaHandle, 
                hive,
                hiveSchemaHandle,
                hiveDates
            )
            new LazyFrame(h)

        // ---------------------------------------------------------
        // Scan IPC (Stream)
        // ---------------------------------------------------------

        /// <summary>
        /// Lazily read Arrow IPC (Feather v2) from a Stream.
        /// </summary>
        /// <remarks>
        /// This reads the stream fully into memory to construct the Lazy execution plan.
        /// </remarks>
        static member ScanIpc(
            stream: System.IO.Stream,
            ?schema: PolarsSchema,
            ?nRows: uint64,
            ?rechunk: bool,
            ?cache: bool,
            ?rowIndexName: string,
            ?rowIndexOffset: uint32,
            ?includePathColumn: string,
            ?hivePartitioning: bool,
            ?hivePartitionSchema: PolarsSchema,
            ?tryParseHiveDates: bool
        ) : LazyFrame =
            
            use ms = new System.IO.MemoryStream()
            stream.CopyTo(ms)
            let bytes = ms.ToArray()

            LazyFrame.ScanIpc(
                bytes,
                ?schema = schema,
                ?nRows = nRows,
                ?rechunk = rechunk,
                ?cache = cache,
                ?rowIndexName = rowIndexName,
                ?rowIndexOffset = rowIndexOffset,
                ?includePathColumn = includePathColumn,
                ?hivePartitioning = hivePartitioning,
                ?hivePartitionSchema = hivePartitionSchema,
                ?tryParseHiveDates = tryParseHiveDates
            )
    type DataFrame with
        /// <summary>
        /// Read an Arrow IPC (Feather v2) file into a DataFrame.
        /// <para>
        /// Note: This method internally uses LazyFrame.ScanIpc and collects the result.
        /// </para>
        /// </summary>
        static member ReadIpc(
            path: string,
            ?columns: seq<string>,
            ?nRows: uint64,
            ?rechunk: bool,
            ?cache: bool,
            ?glob: bool,
            ?rowIndexName: string,
            ?rowIndexOffset: uint32,
            ?includePathColumn: string,
            ?schema: PolarsSchema,
            ?hivePartitioning: bool,
            ?hivePartitionSchema: PolarsSchema,
            ?tryParseHiveDates: bool,
            ?cloudOptions: CloudOptions
        ) : DataFrame =
            
            let mutable lf = LazyFrame.ScanIpc(
                path,
                ?nRows = nRows,
                ?rechunk = rechunk,
                ?cache = cache,
                ?glob = glob,
                ?rowIndexName = rowIndexName,
                ?rowIndexOffset = rowIndexOffset,
                ?includePathColumn = includePathColumn,
                ?schema = schema,
                ?hivePartitioning = hivePartitioning,
                ?hivePartitionSchema = hivePartitionSchema,
                ?tryParseHiveDates = tryParseHiveDates,
                ?cloudOptions = cloudOptions
            )

            match columns with
            | Some cols -> 
                let mutable colList = cols |> Seq.toList
                if not (List.isEmpty colList) then
                    match rowIndexName with
                    | Some rName when not (List.contains rName colList) -> colList <- colList @ [rName]
                    | _ -> ()

                    match includePathColumn with
                    | Some pName when not (List.contains pName colList) -> colList <- colList @ [pName]
                    | _ -> ()

                    use sel = new Selector(PolarsWrapper.SelectorCols(colList |> List.toArray))
                    lf <- lf.Select [sel :> IColumnExpr]
            | None -> ()

            lf.Collect()

        // ---------------------------------------------------------
        // Read IPC (Memory / Bytes)
        // ---------------------------------------------------------

        /// <summary>
        /// Read IPC from in-memory bytes.
        /// </summary>
        static member ReadIpc(
            buffer: byte[],
            ?columns: seq<string>,
            ?nRows: uint64,
            ?rechunk: bool,
            ?cache: bool,
            ?rowIndexName: string,
            ?rowIndexOffset: uint32,
            ?includePathColumn: string,
            ?schema: PolarsSchema,
            ?hivePartitioning: bool,
            ?hivePartitionSchema: PolarsSchema,
            ?tryParseHiveDates: bool
        ) : DataFrame =
            
            let mutable lf = LazyFrame.ScanIpc(
                buffer,
                ?nRows = nRows,
                ?rechunk = rechunk,
                ?cache = cache,
                ?rowIndexName = rowIndexName,
                ?rowIndexOffset = rowIndexOffset,
                ?includePathColumn = includePathColumn,
                ?schema = schema,
                ?hivePartitioning = hivePartitioning,
                ?hivePartitionSchema = hivePartitionSchema,
                ?tryParseHiveDates = tryParseHiveDates
            )

            match columns with
            | Some cols -> 
                let mutable colList = cols |> Seq.toList
                if not (List.isEmpty colList) then
                    match rowIndexName with
                    | Some rName when not (List.contains rName colList) -> colList <- colList @ [rName]
                    | _ -> ()

                    match includePathColumn with
                    | Some pName when not (List.contains pName colList) -> colList <- colList @ [pName]
                    | _ -> ()

                    use sel = new Selector(PolarsWrapper.SelectorCols(colList |> List.toArray))
                    lf <- lf.Select [sel :> IColumnExpr]
            | None -> ()

            lf.Collect()

        // ---------------------------------------------------------
        // Read IPC (Stream)
        // ---------------------------------------------------------

        /// <summary>
        /// Read IPC from a Stream.
        /// </summary>
        static member ReadIpc(
            stream: System.IO.Stream,
            ?columns: seq<string>,
            ?nRows: uint64,
            ?rechunk: bool,
            ?cache: bool,
            ?rowIndexName: string,
            ?rowIndexOffset: uint32,
            ?includePathColumn: string,
            ?schema: PolarsSchema,
            ?hivePartitioning: bool,
            ?hivePartitionSchema: PolarsSchema,
            ?tryParseHiveDates: bool
        ) : DataFrame =
            
            use ms = new System.IO.MemoryStream()
            stream.CopyTo ms
            let bytes = ms.ToArray()

            DataFrame.ReadIpc(
                bytes,
                ?columns = columns,
                ?nRows = nRows,
                ?rechunk = rechunk,
                ?cache = cache,
                ?rowIndexName = rowIndexName,
                ?rowIndexOffset = rowIndexOffset,
                ?includePathColumn = includePathColumn,
                ?schema = schema,
                ?hivePartitioning = hivePartitioning,
                ?hivePartitionSchema = hivePartitionSchema,
                ?tryParseHiveDates = tryParseHiveDates
            )